src/components/Widget/index.tsx

```diff
@@ -51,14 +51,23 @@ export default function Widget({ defaultToken, onTokensChange, onReviewSwapClick
   const { inputs, tokenSelector } = useSyncWidgetInputs(defaultToken)
   const { settings } = useSyncWidgetSettings()
   const { transactions } = useSyncWidgetTransactions()
+
+  const lastValue = useRef([inputs.value.INPUT, inputs.value.OUTPUT])
+  useEffect(() => {
+    const [input, output] = [inputs.value.INPUT, inputs.value.OUTPUT]
+    const [lastInput, lastOutput] = lastValue.current
+    // Avoid calling onTokensChange if only the handler has changed.
+    if (input === lastInput && output === lastOutput) return
+    if (input && lastInput && input.equals(lastInput) && output && lastOutput && lastOutput.equals(lastOutput)) return
```

Comment:
are lines 60 and 61 doing the same thing / kinda redundant? can we just use one of them?

Reply:
L60 covers the case where there are no tokens selected (ie `input = lastInput = output = lastOutput = undefined`).